src/plugins/explore/public/application/utils/state_management/actions/set_dataset/set_dataset.ts

```diff
@@ -40,12 +38,15 @@ export const setDatasetActionCreator = (
   dispatch(clearResults());
   dispatch(setQueryWithHistory(queryStringState));
 
-  dispatch(
-    setQueryWithHistory({
-      ...queryStringState,
-      ...(dataView ? { dataset: dataView.toDataset() } : {}),
-    })
-  );
+  const datasetFromDataView = dataView ? await dataView.toDataset() : undefined;
+  const updatedQuery = {
+    ...queryStringState,
+    ...(datasetFromDataView ? { dataset: datasetFromDataView } : {}),
+  };
+
+  dispatch(setQueryWithHistory(updatedQuery));
```

Comment:
I think sync up with query$ history twice can be a bit problematic? should we consider remove line 39?

Reply:
+1 to anan's question